### PR TITLE
Refactor non-critical code to remove excessive inlining

### PR DESCRIPTION
**Pull request by yfguo**
ch4: code refactoring

The main goal of the refactoring is to remove the inlining for
non-performance-critical functions. For clarification, the
``non-performance-critical'' functions refer to the functions that are
usually not used in mixed with the computation iterations in a MPI
application such as MPI_Init, comm creation, win creation, spawn, etc.
The implementation of these uninlined functions are moved to their
corresponding .c files. Specifically, this refactoring contains the
following changes and cleanups.

* [x] CH4 Level
    1. [x] Uninlined all CH4 level functions that is not performance-critical
       and moved the implementation to .c files. (ch4r_win.h pending)
    2. [x] Remove duplicated declarications for uninlined MPID_* function from
       the .h since they are already in the `mpidch4.h`.
    3. [x] Remove namespace prefix for internal functions (the ones that only
       used in .c files). Add declarications for them to suppress warnings.
    4. [x] Remove unused functions.
    5. [x] Formatted the code as a results of name and attribute changes.

* [ ] NETMOD Level
    * [x] shared parts
        1. [x] `netmod.h` and `netmod_impl.h`
        2. [x] new `netmod_impl.c` for uninlined wrapper functions
    * [ ] OFI
        1. [x] init, comm, spawn, win, op, datatype
        2. [x] impl
        3. [ ] am handling related
    * [ ] UCX, PTL4, STUB
* [ ] SHMMOD  Level
    If any, will look later this week depending the progress of the hackathon.
* [ ] Generic Level
    Later this week.

### CODE REVIEW

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

File: src/mpid/ch4/src/ch4i_comm.c [hunk 1]
```diff
@@ -0,0 +1,1118 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2016 by Argonne National Laboratory.
```
Comment (hajimefu) [inline src/mpid/ch4/src/ch4i_comm.c]: I don't know the policy, but should the newly-created files have the up-to-date year here (2018)?

File: src/mpid/ch4/include/mpidch4.h [hunk 2]
```diff
@@ -18,6 +18,8 @@
  */
 #define MPIDI_CH4I_API(rc,fcnname,...)            \
   MPL_STATIC_INLINE_PREFIX rc MPID_##fcnname(__VA_ARGS__) MPL_STATIC_INLINE_SUFFIX
+#define MPIDI_CH4I_API_NOINLINE(rc,fcnname,...)            \
+  rc MPID_##fcnname(__VA_ARGS__)
```
Comment (hajimefu) [inline src/mpid/ch4/include/mpidch4.h]: I would define this in a separate commit.

File: src/mpid/ch4/src/ch4_init.c [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+
+#include "mpidimpl.h"
+
+/* #include "ch4_impl.h" */
+/* #include "ch4r_proc.h" */
+/* #include "ch4i_comm.h" */
+/* #include "ch4_comm.h" */
+/* #include "strings.h" */
+/* #include "datatype.h" */
+/* #include "ch4r_recvq.h" */
```
Comment (hajimefu) [inline src/mpid/ch4/src/ch4_init.c]: Remove them.

File: src/mpid/ch4/src/ch4r_symheap.h [hunk 4]
```diff
@@ -299,6 +299,7 @@ static inline int MPIDI_CH4R_get_symmetric_heap(MPI_Aint size,
     MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_CH4R_GET_SYMMETRIC_HEAP);
     return mpi_errno;
   fn_fail:
+    *base = NULL;
```
Comment (hajimefu) [inline src/mpid/ch4/src/ch4r_symheap.h]: Why is this included in this patch?

File: src/mpid/ch4/src/ch4r_init.h [hunk 5]
```diff
@@ -11,143 +11,19 @@
...
-    return MPI_SUCCESS;
-}
+// #include "ch4_impl.h"
+// #include "ch4i_util.h"
+// #include "ch4r_buf.h"
+// #include "ch4r_callbacks.h"
+// #include "ch4r_rma_target_callbacks.h"
+// #include "ch4r_rma_origin_callbacks.h"
+// #include "mpidig.h"
+// #include "uthash.h"
```
Comment (hajimefu) [inline src/mpid/ch4/src/ch4r_init.h]: Remove them.

**Comment by hajimefu** [comment]
Wait... the build failure might be my fault. Let me take a look.

**Comment by hajimefu** [comment]
@yfguo Can you pick https://github.com/hajimefu/mpich/commit/7be898296856c22046fb7506bed7dfb1cb5bcf5c ? This should fix the build issue with `no-inline` builds.

**Comment by hajimefu** [comment]
- [ ] Make sure to pick changes in 421a3e03c when rebasing

File: src/mpid/ch4/src/ch4i_comm.c (hunk 1 above)
Comment (yfguo) [reply src/mpid/ch4/src/ch4i_comm.c]: Good point, I will update it.

File: src/mpid/ch4/include/mpidch4.h (hunk 2 above)
Comment (yfguo) [reply src/mpid/ch4/include/mpidch4.h]: I think I will squash the whole thing into one commit for CH4, another for each netmod/shm/generic. Probably leave this in the CH4 one for atomicity.

**Comment by yfguo** [comment]
try ping myself @yfguo

**Comment by hajimefu** [comment]
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by hajimefu** [comment]
test:intel:ch4/ofi/sockets/gnu/all/all/debug,default

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

**Comment by yfguo** [comment]
@hajimefu It turns out that we need to change the function names in the `func_table.c` as well. Otherwise, I will segfault at runtime.

**Comment by yfguo** [comment]
I have merged @gcongiu 's namespace refactoring PR with some changes. The commits need some squash and cleanup, but the aggregation of changes is a namespace scheme that I am satisfied now.

**Comment by yfguo** [comment]
test:jenkins/ch4/ucx

**Comment by yfguo** [comment]
test:jenkins/ch4/ucx
